### PR TITLE
[Bug Fixed] update the cast of ScalingTensor-FP16

### PR DESCRIPTION
**Pull request by wkcn**
**Description**
In the function `cast_to_fp16`, [`input_fp16 = (input * meta.scale).to(torch.float16)`](https://github.com/Azure/MS-AMP/blob/main/msamp/common/tensor/cast.py#L74), if the FP32 scalar `meta.scale` is large than 65504, i.e. the maximum value of FP16, `meta.scale` will be casted to the overflow value, INF by mistake. (Please refer to the PyTorch issue: https://github.com/pytorch/pytorch/issues/98691). It leads to a wrong output.

**Reproduce**
```python
import torch
from msamp.common.tensor import tensor
from msamp.common.dtype import Dtypes

fp16_value = torch.tensor([1.0 / (2 ** 17)], dtype=torch.float16, device='cuda')
fp32_value = fp16_value.float()
for dtype in [torch.float16, torch.float32]:
    print(dtype)
    value = fp16_value.to(dtype)
    scaling_fp16 = value.cast(Dtypes.kfloat16)
    print(torch.allclose(fp32_value, scaling_fp16.float()))
    scaling_fp8 = scaling_fp16.cast(Dtypes.kfloat8_e4m3)
    print(torch.allclose(fp32_value, scaling_fp8.float()))
    print(fp32_value.float().item(), scaling_fp16.float().item(), scaling_fp8.float().item())
```

Output before Bug Fixed:
```
torch.float16
False
False
7.62939453125e-06 inf 1.3351438610698096e-05
torch.float32
True
True
7.62939453125e-06 7.629393621755298e-06 7.629393621755298e-06
```

Output after Bug Fixed: 
```
torch.float16
True
True
7.62939453125e-06 7.62939453125e-06 7.629393621755298e-06
torch.float32
True
True
7.62939453125e-06 7.629395440744702e-06 7.629393621755298e-06
```

**Major Revision**
- support conversion between any two data types (fp8_e4m3, fp8_e5m2, bf16, fp16, fp32 w/ or w/o ScalingTensor).
- re-write the function `cast_to_fp16`
- Add ScalingMeta.in_time_scaling_context
- Unit test

**Minor Revision**
In ScalingMeta, specific FP32 dtype for amax, scale, scale_inv

### CODE REVIEW

File: tests/common/tensor/test_tensor.py [hunk 1]
```diff
@@ -26,39 +26,6 @@ def tearDown(self):
         """Hook method for deconstructing the test fixture after testing it."""
         pass
 
-    @decorator.cuda_test
```
Comment (tocean) [inline tests/common/tensor/test_tensor.py:29]: Removing this UT means we support to cast to kbfloat16?

File: msamp/common/tensor/cast.py [hunk 2]
```diff
@@ -77,22 +77,20 @@ def cast_to_fp16(input, meta, sync=False):
...
-        if in_time or sync:
-            # notice: we scale the tensor with qtype FP8-E4M3.
-            meta.reset_scaling_factor(qtype=Dtypes.kfloat8_e4m3)
-        meta.scale.clamp_(max=Floating.qfp_max[meta.qtype])
+        meta.reset_scaling_factor()
 
         meta.scale_inv.data.copy_(torch.reciprocal(meta.scale))    # scale_inv = 1 / scale
-        input_fp16 = (input * meta.scale).to(torch.float16)
-        return input_fp16
+        dtype = Dtypes.get_dtype_from_qtype(meta.qtype)
```
Comment (tocean) [inline msamp/common/tensor/cast.py:90]: This code change is for supporting bf16?

Comment (wkcn) [reply msamp/common/tensor/cast.py:90]: Yes. It supports bf16/fp16 and fp32.

File: tests/common/tensor/test_tensor.py (hunk 1 above)
Comment (wkcn) [reply tests/common/tensor/test_tensor.py:29]: Yes. The PR supports all conversion (refer to the function `test_tensor_cast`).

**Review by tocean — APPROVED** [review]

File: tests/common/tensor/test_tensor.py [hunk 3]
```diff
@@ -208,17 +191,6 @@ def test_tensor_transpose(self):
         transpose_tensor_value = scaling_tensor.t().contiguous().float()
         self.assertTrue(torch.equal(float_tensor.t(), transpose_tensor_value))
 
-    @decorator.cuda_test
-    def test_tensor_fp8_transpose(self):
```
[the remark this answers is not in the document]
Comment (wkcn) [reply tests/common/tensor/test_tensor.py]: The UT has been back.

File: msamp/common/tensor/tensor.py [hunk 4]
```diff
@@ -219,35 +223,43 @@ def _get_cast_from_fn(self):
     def cast(self, qtype):
         """Cast the ScalingTensor by qtype.
 
-        Currently only supports:
```
Comment (guoshzhao) [inline msamp/common/tensor/tensor.py:222]: Need to keep comments to show the ability to cast any types?

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**